### ISSUE TITLE
Byzantine LSEQ-Node.add problem

### ISSUE THREAD

**Issue by Tavistock** [Chat-Wane/LSEQTree#10]
This may be a "won't fix" but I thought about a byzantine problem arising from just accepting random adds. Someone who wants to mess up your local state or through some bug you may get and add request where the subcounters or children of an add are not singly linked. The add expects singly linked nodes so a solution may be adding a checkLimb function. (I use the semantic of limb for singly linked lseq-nodes)

**Comment by Chat-Wane** [comment]
You are right on this point. The function add expects a singly linked node as argument. It's not a "won't fix". It's more like a "will eventually fix". Indeed, in some case, it would be nice to add a batch of elements. For instance, you are working offline on your document, then, you rejoin the network to work in real time with your team members. Either the changes are sent one-by-one (as it is currently assumed), or you send the changes as a tree to factorize the common parts of the identifiers. It would improve the size of messages travelling through the network, and the insertion algorithm (not in the worst case though).

So the most straightforward solution is checkLimb indeed. But eventually, accept a "tree" as argument is very interesting too.

**Comment by Tavistock** [comment]
I don't know if you could have a merge function (as apposed to an add function) because LSEQ isn't a constantly growing data structure. You could have a special structure that represented adds and deletes that was different from LSEQ that was LSEQ plus tombstones. (correct me if I am wrong)

The two use cases for this would be compressing a ledger and diffing a LSEQ and then sending the changes (as represented as a LSEQ+T) over network for smaller messages like you said.

The ledger idea is that instead of collecting a bunch of operations to commmit to your lseq tree (while you have network adds and removes on hold because your user is typing or something like that) you compress those operations into a lseq+t and then call the merge function.

There may be more problems with LSEQ+T (it may need logical epochs for logically non-monotonic operations... haven't thought too much about it) but I think it is a good way of adding the typical CRDT merge function to LSEQ.

**Comment by Tavistock** [comment]
If this seems right to you (I haven't fully implemented a lseq-tree fully or version vectors yet so it's as best I can reckon with my understanding) we should make it another issue so we can talk about it more.
